Add a `Motivation` section in README to explain the problem solved by CLP IR. (#38)
y-scope/clp-loglib-py: README.md

# CLP Python Logging Library

This is a Python `logging` library meant to supplement [CLP (Compressed Log Processor)][0].
It operates by serializing and compressing log events using the CLP Intermediate Representation (IR)

> format, achieving both data streaming capability and effective compression ratios. Log files

Reviewer: ```suggestion
format, achieving both data streaming capabilities and effective compression ratios. Log files
```

> serialized using IR format can be viewed using [YScope Log Viewer][2]. They can also be deserialized

Reviewer: ```suggestion
serialized using the IR format can be viewed using the [YScope Log Viewer][2]. They can also be deserialized
```

to their original plain-text format, or programmatically analyzed with the APIs provided by
[clp-ffi-py][9]. For further information, refer to the detailed explanation in this [Uber blog][1].

## Motivation

CLP buffers a substantial volume of log files before executing compression for a better compression
ratio. However, most individual log files are actively opened for appending over an extended
duration. In their raw-text format, these log files are not space-efficient and do not support
efficient querying through standard text-based tools like `grep`.

To address this problem, this logging library is designed to serialize log events directly in CLP's

> Intermediate Representation (IR) format. A log event created with a CLP logging handler will be
> first encoded into the IR format, and then appended to a compressed output stream. This approach not

Reviewer 2: ```suggestion
Intermediate Representation (IR) format. A log event created with a CLP logging handler will first
be encoded into the IR format, and then appended to a compressed output stream. This approach not
```

## CLP readers (decoders)

> Important Update: The readers and all the other non-logging APIs currently available in this library

Reviewer 2: Can use GitHub markdown [alerts](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts).

are scheduled for deprecation in the upcoming release. To access our newest and improved CLP IR
analytic interface, which offers advanced features like high-performance decoding and enhanced query

> search capabilities, please check and install [clp-ffi-py][9].

Reviewer 2: ```suggestion
are scheduled for deprecation in an upcoming release. To access our newest and improved CLP IR
analytics interface (which offers advanced features like high-performance decoding and enhanced query
search capabilities) check out [clp-ffi-py][9].
```

## CLP readers (decoders)

> [!WARNING]

> > Important Update: The readers and all the other non-logging APIs currently available in this
> > library are scheduled for deprecation in an upcoming release. To access our newest and improved
> > CLP IR analytics interface (which offers advanced features like high-performance decoding and
> > enhanced query search capabilities) check out [clp-ffi-py][9].

Reviewer 2: ```suggestion
> The readers and all the other non-logging APIs currently available in this library are scheduled
> for deprecation in an upcoming release. To access our newest and improved CLP IR analytics
> interface (which offers advanced features like high-performance decoding and enhanced query search
> capabilities) check out [clp-ffi-py][9].
```

[6]: https://black.readthedocs.io/en/stable/index.html
[7]: https://docformatter.readthedocs.io/en/latest/faq.html#interaction-with-black
[8]: https://beta.ruff.rs/docs/

> [9]: https://github.com/y-scope/clp-ffi-py

Reviewer 2: Finish files with empty lines.
